packages/@lwc/engine-dom/src/renderer/context.ts

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import {
+    createContextProviderWithRegister,
+    WireAdapterConstructor,
+    WireContextValue,
+    WireContextSubscriptionPayload,
+    WireContextSubscriptionCallback,
+} from '@lwc/engine-core';
+import { addEventListener, dispatchEvent } from './index';
+
+export class WireContextSubscriptionEvent extends CustomEvent<undefined> {
+    // These are initialized on the constructor via defineProperties.
+    public readonly setNewContext!: (newContext: WireContextValue) => void;
+    public readonly setDisconnectedCallback!: (disconnectCallback: () => void) => void;
+
+    constructor(
+        adapterToken: string,
+        { setNewContext, setDisconnectedCallback }: WireContextSubscriptionPayload
+    ) {
+        super(adapterToken, {
+            bubbles: true,
+            composed: true,
+        });
+
+        this.setNewContext = setNewContext;
+        this.setDisconnectedCallback = setDisconnectedCallback;
+    }
+}
+
+export function createContextProvider(adapter: WireAdapterConstructor) {
+    return createContextProviderWithRegister(adapter, registerContextProvider);
+}
+
+export function registerContextConsumer(
+    elm: Node,
+    adapterContextToken: string,
+    subscriptionPayload: WireContextSubscriptionPayload
+) {
+    dispatchEvent(elm, new WireContextSubscriptionEvent(adapterContextToken, subscriptionPayload));
+}
+
+export function registerContextProvider(
+    elm: Node,
+    adapterContextToken: string,
+    onContextSubscription: WireContextSubscriptionCallback
+) {
+    addEventListener(elm, adapterContextToken, ((evt: WireContextSubscriptionEvent) => {
+        evt.stopImmediatePropagation();
```

Comment:
Is there a test to confirm that this `stopImmediatePropagation` is doing what we expect it to, and that the behavior is the same on the server and the client?

I.e. It looks like, if there is another context provider higher in the tree, then it will never receive the event because the propagation was stopped at this point. It'd be great to have a test for both the server and client side to confirm that that's the actual behavior.

Reply:
I don't think there is a test for this, and there should be. Thanks for the call out.

Reply:
I was wrong, [there is a test](https://github.com/salesforce/lwc/blob/3cb07f92c7b97c348db799665fda3efe9408678a/packages/@lwc/integration-karma/test/context/simple-context.spec.js#L69-L85).

Reply:
Modified the `context-nested` test to incorporate context "shadowing" and validate that the closest provider is the one providing the value.